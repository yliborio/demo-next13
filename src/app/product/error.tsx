"use client";

export default function ErrorPage() {
  return <h1> Product not found </h1>;
}
