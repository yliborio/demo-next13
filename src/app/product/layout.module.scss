.container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 1rem;
  margin: 1rem;
  .back {
    width: fit-content;
  }
}
