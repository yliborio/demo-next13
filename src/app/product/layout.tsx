import Link from "next/link";
import styles from "./layout.module.scss";
import { BackIcon } from "core/components/icons/back-button";
import { Suspense } from "react";
import Loading from "./loading";

export default function Layout({ children }: { children: React.ReactNode }) {
  return (
    <div className={styles["container"]}>
      <Link
        className={styles["back"]}
        href={"/"}
        aria-label="back to home page"
      >
        <BackIcon />
      </Link>
      <Suspense fallback={<Loading />}>{children}</Suspense>
    </div>
  );
}
