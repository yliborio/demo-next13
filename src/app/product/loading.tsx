import Skeleton from "react-loading-skeleton";
import "react-loading-skeleton/dist/skeleton.css";

export default function Loading() {
  return (
    <div
      style={{
        padding: "2rem 5rem",
      }}
    >
      <Skeleton
        height={"300px"}
        width={"100%"}
        count={1}
        baseColor="darkgrey"
      />
    </div>
  );
}
