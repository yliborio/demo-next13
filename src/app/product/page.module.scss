.container {
  display: flex;
  flex-direction: column;
  width: 30vw;
  align-items: center;
  justify-content: center;
  .title {
    font-size: 20px;
    line-height: 18px;
    margin-bottom: 1rem;
  }
  .action {
    display: flex;
    flex-direction: row;
    gap: 1rem;
    .input {
      height: 2rem;
    }
    .button {
      background-color: black;
      width: 25%;
    }
  }
}
