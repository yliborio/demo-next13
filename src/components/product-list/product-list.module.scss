@import "../../styles/variables.scss";
.container {
  display: flex;
  flex-direction: column;
  gap: 3rem;
  justify-content: center;
  align-items: center;
  width: 100%;
  .products-grid {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));

    grid-gap: 2rem;
    width: 100%;
    .empty {
      display: flex;
      justify-content: center;
      align-items: center;
      font-size: $font-size-2;
    }
  }

  .selector {
    display: flex;
    flex-direction: row;
    gap: 5px;
    font-size: $font-size-2;
    > button {
      border: none;
    }

    .pages {
      &:disabled {
        background-color: $bg-secondary;
        cursor: auto;
      }
      color: black;
      display: flex;
      background-color: transparent;
      font-size: $font-size-2;
      align-items: center;
      justify-content: center;

      width: 30px;
      height: 30px;
      border-radius: 8px;
      cursor: pointer;
    }

    .actions {
      &:disabled {
        cursor: auto;
      }
      font-size: $font-size-2;
      font-weight: 600;
      cursor: pointer;
      border: none;
      background-color: transparent;
      width: 30px;
      height: 30px;
      border-radius: 100%;
    }
  }
}
