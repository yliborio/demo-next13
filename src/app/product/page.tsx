import { ProductInfo } from "core/components/product-info/product-info";
import { FakeAPIProduct } from "core/types/product";

interface PageProps {
  searchParams: { id: string };
}

export default async function Page(props: PageProps) {
  const {
    searchParams: { id },
  } = props;
  const data = await fetch(`https://fakestoreapi.com/products/${id}`);
  const product: FakeAPIProduct = await data.json();

  return <ProductInfo product={product} />;
}
