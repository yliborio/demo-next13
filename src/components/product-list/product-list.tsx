import { FakeAPIProduct } from "core/types/product";
import { ProductCard } from "../product-card/product-card";
import styles from "./product-list.module.scss";

interface ProductListProps {
  products: FakeAPIProduct[];
}

export const ProductList = ({ products }: ProductListProps) => (
  <div className={styles["container"]}>
    {products?.map((product) => (
      <ProductCard key={product.id} product={product} />
    ))}
  </div>
);
