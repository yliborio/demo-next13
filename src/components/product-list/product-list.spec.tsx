import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { ProductList } from "./product-list";
import { FakeAPIProduct } from "core/types/product";

describe("ProductList", () => {
  it("displays a list of products correctly", () => {
    const products: FakeAPIProduct[] = [
      {
        id: 1,
        title: "Product 1",
        price: 19.99,
        image: "/sample-image-test",
        rating: {
          rate: 4,
          count: 4,
        },
      } as any,
      {
        id: 2,
        title: "Product 2",
        price: 24.99,
        image: "/sample-image-test",
        rating: {
          rate: 4,
          count: 4,
        },
      },
    ];

    const { asFragment } = render(<ProductList products={products} />);
    expect(asFragment()).toMatchSnapshot();

    expect(screen.getByText("Product 1 - 4/5 (4)")).toBeInTheDocument();
    expect(screen.getByText("Product 2 - 4/5 (4)")).toBeInTheDocument();
    expect(screen.getAllByRole("link")).toHaveLength(products.length);
  });

  it("should display n items per page", () => {
    jest.spyOn(window, "scrollTo").mockImplementation(jest.fn());
    const products: FakeAPIProduct[] = [
      {
        id: 1,
        title: "Product 1",
        price: 19.99,
        image: "/sample-image-test",
        rating: {
          rate: 4,
          count: 4,
        },
      } as any,
      {
        id: 2,
        title: "Product 2",
        price: 24.99,
        image: "/sample-image-test",
        rating: {
          rate: 4,
          count: 4,
        },
      },
      {
        id: 3,
        title: "Product 3",
        price: 24.99,
        image: "/sample-image-test",
        rating: {
          rate: 4,
          count: 4,
        },
      },
    ];

    render(<ProductList products={products} items={1} />);

    expect(screen.queryByText("Product 1 - 4/5 (4)")).toBeInTheDocument();
    expect(screen.queryByText("Product 2 - 4/5 (4)")).toBeNull();
    expect(screen.queryByText("Product 3 - 4/5 (4)")).toBeNull();

    // back button should be disabled in first page
    expect(screen.getByTestId("back-button")).toBeDisabled();

    // go to next page
    fireEvent.click(screen.getByTestId("next-button"));

    //only item 2 is presented
    expect(screen.queryByText("Product 1 - 4/5 (4)")).toBeNull();
    expect(screen.queryByText("Product 2 - 4/5 (4)")).toBeInTheDocument();
    expect(screen.queryByText("Product 3 - 4/5 (4)")).toBeNull();

    // go back
    fireEvent.click(screen.getByTestId("back-button"));
    expect(screen.queryByText("Product 1 - 4/5 (4)")).toBeInTheDocument();

    fireEvent.click(screen.getByTestId("page-2-button"));
    expect(screen.queryByText("Product 3 - 4/5 (4)")).toBeInTheDocument();

    // next button should be disabled in last page
    expect(screen.getByTestId("next-button")).toBeDisabled();
  });

  it('displays a "No products found" message when no products are available', () => {
    const products: FakeAPIProduct[] = [];

    render(<ProductList products={products} />);
    expect(screen.getByText("No products found :(")).toBeInTheDocument();
  });
});
