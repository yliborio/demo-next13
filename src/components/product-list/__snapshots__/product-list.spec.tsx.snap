// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList displays a list of products correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="products-grid"
    >
      <a
        class="container"
        href="/product?id=1"
      >
        <div
          class="card"
        >
          <div
            class="product-image"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="0"
              loading="lazy"
              sizes="100vw"
              src="/_next/image?url=%2Fsample-image-test&w=3840&q=75"
              srcset="/_next/image?url=%2Fsample-image-test&w=640&q=75 640w, /_next/image?url=%2Fsample-image-test&w=750&q=75 750w, /_next/image?url=%2Fsample-image-test&w=828&q=75 828w, /_next/image?url=%2Fsample-image-test&w=1080&q=75 1080w, /_next/image?url=%2Fsample-image-test&w=1200&q=75 1200w, /_next/image?url=%2Fsample-image-test&w=1920&q=75 1920w, /_next/image?url=%2Fsample-image-test&w=2048&q=75 2048w, /_next/image?url=%2Fsample-image-test&w=3840&q=75 3840w"
              style="color: transparent;"
              width="0"
            />
          </div>
          <div
            class="title"
          >
            <span>
              Product 1 - 4/5  (4)
            </span>
          </div>
          <span
            class="divider"
          />
          <span
            class="price"
          >
            $19.99 
          </span>
        </div>
      </a>
      <a
        class="container"
        href="/product?id=2"
      >
        <div
          class="card"
        >
          <div
            class="product-image"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="0"
              loading="lazy"
              sizes="100vw"
              src="/_next/image?url=%2Fsample-image-test&w=3840&q=75"
              srcset="/_next/image?url=%2Fsample-image-test&w=640&q=75 640w, /_next/image?url=%2Fsample-image-test&w=750&q=75 750w, /_next/image?url=%2Fsample-image-test&w=828&q=75 828w, /_next/image?url=%2Fsample-image-test&w=1080&q=75 1080w, /_next/image?url=%2Fsample-image-test&w=1200&q=75 1200w, /_next/image?url=%2Fsample-image-test&w=1920&q=75 1920w, /_next/image?url=%2Fsample-image-test&w=2048&q=75 2048w, /_next/image?url=%2Fsample-image-test&w=3840&q=75 3840w"
              style="color: transparent;"
              width="0"
            />
          </div>
          <div
            class="title"
          >
            <span>
              Product 2 - 4/5  (4)
            </span>
          </div>
          <span
            class="divider"
          />
          <span
            class="price"
          >
            $24.99 
          </span>
        </div>
      </a>
    </div>
    <div
      class="selector"
    >
      <button
        class="actions"
        data-testid="back-button"
        disabled=""
      >
        &lt;
      </button>
      <button
        class="pages"
        data-testid="page-0-button"
        disabled=""
        tabindex="0"
      >
        1
      </button>
      <button
        class="actions"
        data-testid="next-button"
        disabled=""
      >
        &gt;
      </button>
    </div>
  </div>
</DocumentFragment>
`;
